Extend IEEE 802.15.4 retransmissions support

jciupis opened the pull request.

Commit:
net: pkt: add IEEE 802.15.4 security-related flags

This commit extends the struct net_pkt structure to contain flags
related to IEEE 802.15.4 security processing. Added flags:
 * ieee802154_frame_secured - indicates if a frame is authenticated and
encrypted according to the configuration stored in Auxiliary Security
Header
 * ieee802154_mac_hdr_rdy - indicates if MAC header is ready to be
transmitted or if it requires further modifications

These flags can be used by the upper layer to correctly configure
retransmissions of frames.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

Commit:
drivers: ieee802154: align with API changes

This commit aligns the nRF5 ieee802154 driver with the latest
API changes necessary to handle security-related flags properly.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

Commit:
manifest: update hal_nordic revision

This commit updates hal_nordic revision to bring the latest changes in
the nRF IEEE 802.15.4 radio driver.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

hubertmis commented on include/net/net_pkt.h:
```diff
@@ -240,6 +240,12 @@ struct net_pkt {
 	uint8_t ieee802154_arb : 1; /* ACK Request Bit is set in the frame */
 	uint8_t ieee802154_ack_fpb : 1; /* Frame Pending Bit was set in the ACK */
 	uint8_t ieee802154_frame_retry : 1; /* The frame is being retransmitted. */
```
I thought the new parameters are to replace this one. Is there a need to keep `ieee_frame_retry` with more detailed data about frame counter, IEs, and if the frame is secured?

jciupis replied:
From my point of view this flag can be removed.

hubertmis commented on include/net/net_pkt.h:
```diff
@@ -240,6 +240,12 @@ struct net_pkt {
+					       * Auxiliary Security Header
+					       */
+	uint8_t ieee802154_fc_is_set : 1; /* Frame Counter was set */
+	uint8_t ieee802154_ie_is_updated : 1; /* Information Elements were updated */
```
I would extend this comment a little to indicate that only selected IEs can be updated by radio. The list includes IEs with precise timing data unknown by the upper layer. I think the only one so far is CSL IE.

jciupis replied:
Please check if the current doc is accurate enough.

hubertmis replied:
Perfect!

Commit:
net: pkt: remove ieee802154_frame_retry flag

Flag that indicates if a given packet is being retransmitted has become
obsolete since more detailed flags were added. This commit removes the
flag and references to it altogether.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

edmont commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -433,42 +433,75 @@ static void nrf5_tx_started(const struct device *dev,
 
+static bool nrf5_tx_immediate(struct net_pkt *pkt, uint8_t *payload, bool cca)
+{
+	if ((pkt->ieee802154_fc_is_set && (!pkt->ieee802154_ie_is_updated)) ||
```
This is equivalent to: `if (pkt->ieee802154_fc_is_set  != pkt->ieee802154_ie_is_updated) {`

However it's probably being optimized out by the compiler. But for the current state of the radio driver APIs it can be moved safely from all the three transmission functions to `nrf5_tx` to reduce the code size.

jciupis replied:
That's a very good catch, thank you.

edmont commented on include/net/net_pkt.h:
```diff
@@ -239,7 +239,16 @@ struct net_pkt {
+					       * Auxiliary Security Header
+					       */
+	uint8_t ieee802154_fc_is_set : 1; /* Frame Counter was set */
+	uint8_t ieee802154_ie_is_updated : 1; /* Information Elements that contain
```
Should this one be conditioned to `CONFIG_IEEE802154_2015`?

edmont reviewed (approved):
LGTM, thanks.

rlubos reviewed (approved):
LGTM

rlubos commented on west.yml:
```diff
@@ -62,7 +62,7 @@ manifest:
       path: modules/hal/microchip
     - name: hal_nordic
-      revision: 74b3b21f60aa3dc9a4364ffc28dbb47ad8b699a9
+      revision: pull/100/head
```
Please update SHA: `d979c8dc313e4dc3e52d5606c28622e5278be50c`

jciupis replied:
Done.

pfalcon reviewed (approved):
As this includes changes to include/net/net_pkt.h, adding @jukkar to reviewers.

However, +1'ing myself based on other reviews.
